packages/ai/src/core.ts

```diff
@@ -68,37 +69,68 @@ export class EdgeDBAI {
     });
   }
 
-  private async fetchRag(request: Omit<RAGRequest, "model" | "prompt">) {
+  private async fetchRag(request: RagRequest, context: QueryContext) {
     const headers = request.stream
       ? { Accept: "text/event-stream", "Content-Type": "application/json" }
       : { Accept: "application/json", "Content-Type": "application/json" };
 
+    const { messages: initialMessages, prompt } = request;
+
+    if (prompt && initialMessages)
+      throw new Error(
+        "You can provide either a prompt or a messages array, not both.",
+      );
+
+    const messages: EdgeDBMessage[] | undefined = prompt
+      ? [{ role: "user" as const, content: [{ type: "text", text: prompt }] }]
+      : initialMessages;
+
+    const providedPrompt =
+      this.options.prompt &&
+      ("name" in this.options.prompt || "id" in this.options.prompt);
+
     const response = await (
       await this.authenticatedFetch
     )("rag", {
       method: "POST",
       headers,
       body: JSON.stringify({
         ...request,
+        context,
         model: this.options.model,
-        prompt: this.options.prompt,
+        prompt: {
+          ...this.options.prompt,
+          // if user provides prompt.custom without id/name it is his choice
+          // to not include default prompt msgs, but if user provides messages
+          // and doesn't provide prompt.custom, since we add messages to the
+          // prompt.custom we also have to include default prompt messages
+          ...(!this.options.prompt?.custom &&
+            !providedPrompt && {
+              name: "builtin::rag-default",
+            }),
+          custom: [...(this.options.prompt?.custom || []), ...messages!],
+        },
+        query: [...messages!].reverse().find((msg) => msg.role === "user")!
```

Comment:
Can we make `messages` an easier type to deal with earlier? Do we lose anything to add `?? []` at the end of the initial assignment to get `EdgeDBMessage[]`? Maybe there is a semantic difference between `undefined` and `[]` that we want to preserve?